R/adjustcarryforward.R

```diff
@@ -302,366 +729,380 @@ adjustcarryforward <- function(subjid,
-
-        if (num.exclude > 1) {
-          # first order by decreasing temp.diff (where rep=T)
-          worst.row <- order(rep, df$temp.diff, decreasing = T)[1]
-          df[worst.row, exclude := temp.exclude]
-        }
-
-        return(df$exclude)
-
-
-      })(copy(.SD))]
-
-
+      # "include" protects subsets (used in option 3)
+      subj.df[!grepl("Exclude", exclude) & !grepl("Include", exclude),
+              exclude := (function (df) {
+                # do steps 15a - 15q (functionalized for ease)
+                eval_df <- calc_temp_exclusion_15(
+                  copy(df),
+                  ewma.fields, tanner.ht.vel, who.ht.vel, exclude.levels,
+                  ewma.exp, minfactor, maxfactor, banddiff, banddiff_plus,
+                  min_ht.exp_under, min_ht.exp_over, max_ht.exp_under, max_ht.exp_over)
+
+                # r.  If there is only one potential exclusion identified in step 15j for a subject and parameter,
+                #     replace exc_ht=15 for that value if it met criteria i, ii, v, or vi  and exc_ht=16 if it met criteria iii, iv, vii, or viii
+                # NOTE: these exclusions are assigned in the code above as 'Exclude-Min-Height-Change' and 'Exclude-Max-Height-Change'
+
+                # count the amount of error codes we get
+                all_exclude = !is.na(eval_df$temp.exclude)
+                num.exclude = sum(all_exclude)
+
+                # if there's only one, all of the groups agree
+                # s.  If there is more than one potential exclusion identified in step 14h for a subject and parameter, determine which value has the largest temp_diff and
+                #     replace exc_ht=15 for that value if it met criteria i, ii, v, or vi and exc_ht=16 for that value if it met criteria iii,  iv, vii, or viii.
+
+                if (exclude_opt == 0){
+                  # option 0: normal
+                  if (num.exclude == 1){
+                    eval_df[all_exclude, exclude := temp.exclude]
+                  } else if (num.exclude > 1) {
+                    # first order by decreasing temp.diff (where rep=T)
+                    worst.row = order(all_exclude, eval_df$temp.diff, decreasing = T)[1]
+                    eval_df[worst.row, exclude := temp.exclude]
+                  }
+                } else if (exclude_opt == 1){
+                  # option 1: if there's a carried forward after, we want to exclude that too, until the next include
+                  if (num.exclude == 1){
+                    wh_exclude <- which(all_exclude)
+
+                    # if it's in a string of carried forwards
+                    if (
+                      eval_df[wh_exclude, "orig.exclude"] == "Exclude-Carried-Forward" &&
+                      wh_exclude != nrow(eval_df) &&
+                      eval_df[wh_exclude+1, "orig.exclude"] == "Exclude-Carried-Forward"
+                    ){
+                      # find the next include OR the row end
```

Comment:
Does "the row end" here mean "the end of the CF string"?

Reply:
Sorry, no -- it means the end of the subject's df, if there's no Include after the CF string.